src/zepben/evolve/streaming/get/network_consumer.py

```diff
@@ -191,50 +191,48 @@ async def _retrieve_network(self) -> GrpcResult[Union[NetworkResult, Exception]]
             return result
 
         hierarchy: NetworkHierarchy = result.result
-        for mrid, gr in hierarchy.geographical_regions.items():
+        for mrid in hierarchy.geographical_regions.keys():
             gr_result = await self._get_identified_object(service, mrid)
             if gr_result.was_failure:
                 return gr_result
 
-        for mrid, sgr in hierarchy.sub_geographical_regions.items():
+        for mrid in hierarchy.sub_geographical_regions.keys():
             sgr_result = await self._get_identified_object(service, mrid)
             if sgr_result.was_failure:
                 return sgr_result
 
-        for mrid, substation in hierarchy.substations.items():
+        for mrid in hierarchy.substations.keys():
             substation_result = await self._get_identified_object(service, mrid)
             if substation_result.was_failure:
                 return substation_result
 
-        for mrid, feeder in hierarchy.feeders.items():
+        for mrid in hierarchy.feeders.keys():
             feeder_result = await self._get_identified_object(service, mrid)
             if feeder_result.was_failure:
                 return feeder_result
 
-        last_num_unresolved = MAX_64_BIT_INTEGER
+        failed = set()
```

Comment:
failed still needs to be reset on each iteration of the loop so that it only ever contains failures from the last cycle of resolving references. (if resolving an unresolved reference fails, the unresolved reference will still be present in the service, so only need the latest round of failures)

you could instead make this `failed = None` and keep the `failed = set()` inside the loop, then your return would work.

we could also consider making failed in `NetworkResult` a `Set[str]` with a default of an empty set, rather than an `Optional[Set[str]]`, which I think would make the result easier to use as the user could avoid none checks.
it would mean your return statement here would have to be along the lines of:

    return GrpcResult(NetworkResult(service)) if failed is None else GrpcResult(NetworkResult(service, failed))

Reply:
Initialising `failed = set()` before and within the 'resolve references' loop.